scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/SoftKeywords.scala

```diff
@@ -0,0 +1,87 @@
+package scala.meta.internal.parsers
+
+import scala.meta.Dialect
+import scala.meta.tokens.Token
+import scala.meta.tokens.Token.Ident
+import scala.meta.internal.classifiers.classifier
+
+class SoftKeywords(dialect: Dialect) {
+
+  def isIdentAnd(token: Token, pred: String => Boolean): Boolean = token match {
+    case Ident(value) if pred(value.stripPrefix("`").stripSuffix("`")) => true
+    case _ => false
+  }
+
+  def isSoftKw(token: Token, name: String): Boolean = isIdentAnd(token, _ == name)
+
+  @classifier
+  trait SkAs {
+    val name = "as"
+    def unapply(token: Token): Boolean = {
+      isSoftKw(token, name) && dialect.allowAsPatternBinding
```

Comment:
Nit: can we move `dialect.allowAsPatternBinding` before the `isSoftKw`? The `isSoftKw` test is a lot more expensive and will be `false` in most cases

Reply:
Sure! Fun fact, this was added with M1 and will actually be removed with M3.

Reply:
Used the advice from @kitbellew and now we have a matches method that first always checks if a particular dialect feature is supported.